File: TUF A2Z DSA Sheet/Sorting/quick_sort.cpp
// Program: Implementation of Quick Sort Algorithm.

#include <bits/stdc++.h>
using namespace std;

int partition(vector<int> &arr, int low, int high) {
    
    int pivot = arr[low];
    int i = low;
    int j = high;

    while(i < j) {
        while(arr[i] <= pivot and i <= high-1) i++;
        while(arr[j] > pivot and j >= low+1) j--;
        if(i < j ) swap(arr[i], arr[j]);
    }
    swap(arr[low], arr[j]);
    return j;
    
}


int quick_sort(vector<int> &arr, int low, int high) {
    
    if(low < high) {
        int pIndex = partition(arr, low, high);
        quick_sort(arr, low, pIndex-1);
        quick_sort(arr, pIndex+1, high);
    }
    
}

int main() {
    
    int n;
    cin >> n;
    vector<int> arr(n); 
    for(int i=0; i<n; i++) {
        cin >> arr[i];
    }

    quick_sort(arr, 0, n-1);

    for(int i : arr) {
        cout << i << " ";
    }

    return 0;
}

File: DSA Implementation in Go/Divide and Conquer/quick_sort.go
//go:build ignore

package main

import "fmt"

// QuickSort sorts an array using the quick sort algorithm
func QuickSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	pivot := arr[len(arr)/2]
	left := []int{}
	right := []int{}

	for i := 0; i < len(arr); i++ {
		if i == len(arr)/2 {
			continue
		}
		if arr[i] < pivot {
			left = append(left, arr[i])
		} else {
			right = append(right, arr[i])
		}
	}

	left = QuickSort(left)
	right = QuickSort(right)

	return append(append(left, pivot), right...)
}

// main function to demonstrate quick sort
func main() {
	arr := []int{38, 27, 43, 3, 9, 82, 10}
	fmt.Print("Unsorted array: ")
	fmt.Println(arr)

	sorted := QuickSort(arr)
	fmt.Print("Sorted array: ")
	fmt.Println(sorted)
}
